ios/RNSnowplowTracker.m

```diff
@@ -14,175 +14,216 @@ - (dispatch_queue_t)methodQueue
+        [builder setLifecycleEvents:[options[@"setLifecycleEvents"] boolValue]];
+        [builder setScreenContext:[options[@"setScreenContext"] boolValue]];
+        [builder setInstallEvent:[options[@"setInstallEvent"] boolValue]];
         [builder setSubject:subject];
+        [builder setSessionContext:[options[@"setSessionContext"] boolValue]];
+        [builder setCheckInterval:[options[@"checkInterval"] integerValue]];
+        [builder setForegroundTimeout:[options[@"foregroundTimeout"] integerValue]];
+        [builder setBackgroundTimeout:[options[@"backgroundTimeout"] integerValue]];
     }];
+
+    if (self.tracker) {
+        resolve(@YES);
+    } else {
+        NSError * error = [NSError errorWithDomain:@"SnowplowTracker" code:200 userInfo:nil];
+        reject(@"ERROR", @"SnowplowTracker: initialize() method - tracker initialisation failed", error);
+    }
 }
 
-RCT_EXPORT_METHOD(setSubjectData :(NSDictionary *)options) {
-      if (options[@"userId"] != nil) {
-              [self.self.tracker.subject setUserId:options[@"userId"]];
-      }
-      if (options[@"screenWidth"] != nil && options[@"screenHeight"] != nil) {
-          [self.self.tracker.subject setResolutionWithWidth:[options[@"screenWidth"] integerValue] andHeight:[options[@"screenHeight"] integerValue]];
-      }
-      if (options[@"viewportWidth"] != nil && options[@"viewportHeight"] != nil) {
-          [self.self.tracker.subject setViewPortWithWidth:[options[@"viewportWidth"] integerValue] andHeight:[options[@"viewportHeight"] integerValue]];
-      }
-      if (options[@"colorDepth"] != nil) {
-          [self.self.tracker.subject setColorDepth:[options[@"colorDepth"] integerValue]];
-      }
-      if (options[@"timezone"] != nil) {
-          [self.self.tracker.subject setTimezone:options[@"timezone"]];
-      }
-      if (options[@"language"] != nil) {
-          [self.self.tracker.subject setLanguage:options[@"language"]];
-      }
-      if (options[@"ipAddress"] != nil) {
-          [self.self.tracker.subject setIpAddress:options[@"ipAddress"]];
-      }
-      if (options[@"useragent"] != nil) {
-          [self.self.tracker.subject setUseragent:options[@"useragent"]];
-      }
-      if (options[@"networkUserId"] != nil) {
-          [self.self.tracker.subject setNetworkUserId:options[@"networkUserId"]];
-      }
-      if (options[@"domainUserId"] != nil) {
-          [self.self.tracker.subject setDomainUserId:options[@"domainUserId"]];
-      }
+RCT_EXPORT_METHOD(setSubjectData :(NSDictionary *)options
+                                :rejecter:(RCTPromiseRejectBlock)reject) {
+
+    // ios throws if it finds an NSNull here - unset the field if null passed
+    if (options[@"userId"] != nil) {
+        if (options[@"userId"] == (id)[NSNull null]) {
+            [self.tracker.subject setUserId:nil];
+        } else {
+            [self.tracker.subject setUserId:options[@"userId"]];
+        }
+    }
```

Comment:
All this could be also written as:
```suggestion
    NSString *userId = options[@"userId"];  // to limit number of access to dictionary
    if (userId) {
        NSString *newUserId = userId == [NSNull null] ? nil : userId;  // No cast to (id)
        [self.tracker.subject setUserId:newUserId];
    }
```